File: Python/level2/괄호 변환.py
def solution(p):
    if not p:
        return p
    
    flag = True
    count = 0
    index = 0
    for i in range(len(p)):
        count += -1 if p[i] == '(' else 1
        if count > 0:
            flag = False
        if count == 0:
            [u, v] = [p[:i + 1], p[i + 1:]]
            if flag:
                return u + solution(v)
            answer = '(' + solution(v) + ')'
            answer += ''.join(['(' if s == ')' else ')' for s in u[1:-1]])
            return answer


# 첫 풀이
# def solution(p):
#     if not p:
#         return p
    
#     count = 0
#     index = 0
#     for i in range(len(p)):
#         count += -1 if p[i] == '(' else 1
#         if count == 0:
#             index = i
#             break
#     [u, v] = [p[:index + 1], p[index + 1:]]
    
#     if isRight(u):
#         return u + solution(v)
#     answer = '(' + solution(v) + ')'
#     answer += ''.join(['(' if s == ')' else ')' for s in u[1:-1]])
#     return answer

# def isRight(string):
#     stack = [string[0]]
#     for i in range(1,len(string)):
#         if len(stack) == 0:
#             stack.append(string[i])
#         elif stack[-1] == '(' and string[i] == ')':
#             stack.pop()
#         else:
#             stack.append(string[i])
#     return True if len(stack) == 0 else False

File: Cplusplus/LEVEL 2/괄호 변환.cpp
//22.02.01 °ýÈ£ º¯È¯ by CJHofficial

#include <string>
#include <vector>
using namespace std;

bool isRight(string target) {
    int check = 0;
    for (int i = 0; i < target.size(); i++) {
        if (target[i] == '(') check += 1;
        else if (target[i] == ')') check -= 1;
        if (check < 0) return false;
    }
    return true;
}

string dfs(string w) {
    if (w == "") return "";
    string u = "";
    string v = "";
    int count = 0;
    int vindex = w.size();
    for (int i = 0; i < w.size(); i++) {
        u += w[i];
        if (w[i] == '(') count += 1;
        if (w[i] == ')') count -= 1;
        if (count == 0) {
            vindex = i + 1;
            break;
        }
    }
    for (int j = vindex; j < w.size(); j++) v += w[j];
    if (isRight(u)) {
        u += dfs(v);
        return u;
    }
    else {
        string temp = "";
        temp += "(";
        temp += dfs(v);
        temp += ")";
        for (int i = 1; i < u.size() - 1; i++) {
            temp += (u[i] == '(') ? ')' : '(';
        }
        return temp;
    }
}
string solution(string p) {
    string answer = "";
    answer = dfs(p);
    return answer;
}